File: react-front/src/types/debug.ts
// Mirrors fastApi-app/schemas/debug.py and the query params on
// fastApi-app/api/debug_vision.py. Used only by the Pipeline Debug screen —
// kept separate from types/api.ts, which is session-scoped.

export interface DebugCheckResult {
  name: string;
  passed: boolean;
  score: number | null;
  message: string;
}

export interface DebugValidationResponse {
  ok: boolean;
  technical_ok: boolean;
  content_ok: boolean | null;
  technical: DebugCheckResult[];
  content: DebugCheckResult[];
  content_skipped_reason: string | null;
  elapsed_ms: number;
}

export type DepthStrategy = "anything" | "blended" | "enhanced_edge";
export type DepthColormap = "none" | "inferno" | "magma" | "turbo" | "jet";
export type SamSource = "depth" | "rgb";
export type NormalHubModel =
  | "metric3d_vit_small"
  | "metric3d_vit_large"
  | "metric3d_vit_giant2";

export interface DepthMapOptions {
  strategy: DepthStrategy;
  model: string;
  colormap: DepthColormap;
}

export interface NormalMapOptions {
  hubModel: NormalHubModel;
}

export interface SamEverythingOptions {
  source: SamSource;
  depthStrategy: DepthStrategy;
  depthModel: string;
  pointsPerSide: number;
  predIouThresh: number;
  stabilityScoreThresh: number;
  minMaskRegionArea: number;
  alpha: number;
}

// A rendered PNG result from either debug image endpoint. `objectUrl` is
// created via URL.createObjectURL — callers own revoking it.
export interface DebugImageResult {
  objectUrl: string;
  elapsedMs: number | null;
  maskCount: number | null;
}

export interface DebugMaskCandidate {
  index: number;
  score: number;
  reason: string;
  clip_checks: Record<string, number> | null;
  preview_b64: string;
  clip_crop_b64: string | null;
  cutout_b64: string;
}

export interface DebugAutoMaskPickResponse {
  click_xy: [number, number] | number[];
  threshold: number;
  winner_index: number | null;
  finalist_indices: number[];
  tiebreak_method: string;
  tiebreak_reason: string | null;
  candidates: DebugMaskCandidate[];
  elapsed_ms: number;
}

export interface DebugSdParams {
  prompt: string;
  negative_prompt: string;
  strength: number;
  num_inference_steps: number;
  guidance_scale: number;
  mask_dilate_pixels?: number;
  compose_dilate_pixels?: number;
}

export interface DebugInpaintAttempt {
  attempt_index: number;
  ok: boolean;
  sd_skipped: boolean;
  scores: Record<string, number>;
  winner_label: string;
  params: DebugSdParams;
  param_fixes_json: string;
  mask_dilate_pixels?: number;
  compose_dilate_pixels?: number;
  mask_pixel_count?: number;
  next_params?: DebugSdParams | null;
  candidate_b64: string;
  clip_crop_b64: string;
  verify_original_crop_b64?: string | null;
}

export interface DebugInpaintVerifyResponse {
  click_xy: [number, number] | number[];
  mask_index: number;
  preview_b64: string;
  cutout_b64: string;
  passed: boolean;
  retries_exhausted: boolean;
  lama_b64: string | null;
  final_b64: string;
  attempts: DebugInpaintAttempt[];
  elapsed_ms: number;
}

File: fastApi-app/schemas/debug.py
from __future__ import annotations

from typing import Annotated

from pydantic import BaseModel, Field


class DebugCheckResult(BaseModel):
    """One technical or content validation check's outcome, for the debug report."""

    name: Annotated[str, Field(description="Check identifier, e.g. 'blur' or 'exposure'.")]
    passed: Annotated[bool, Field(description="Whether this check passed.")]
    score: Annotated[
        float | None, Field(default=None, description="Check-specific numeric score, if any.")
    ]
    message: Annotated[
        str, Field(default="", description="Human-readable detail, populated mainly on failure.")
    ]


class DebugValidationResponse(BaseModel):
    """Full validation scoreboard for one uploaded image — technical + content checks.

    Unlike `POST /images/upload`, this never stops at the first failed check
    and never persists anything; it always returns 200 so a failed check reads
    as data, not an error.
    """

    ok: Annotated[bool, Field(description="True iff every technical and content check passed.")]
    technical_ok: Annotated[bool, Field(description="True iff every technical check passed.")]
    content_ok: Annotated[
        bool | None,
        Field(description="True/False if content validation ran; None if it was skipped."),
    ]
    technical: Annotated[
        list[DebugCheckResult], Field(description="Every technical check, in run order.")
    ]
    content: Annotated[
        list[DebugCheckResult],
        Field(default_factory=list, description="Every content (CLIP) check, if it ran."),
    ]
    content_skipped_reason: Annotated[
        str | None,
        Field(default=None, description="Why the content stage did not run, if it didn't."),
    ]
    elapsed_ms: Annotated[float, Field(description="Total wall time for both stages.")]


class DebugMaskCandidate(BaseModel):
    """One SAM cutout scored by auto mask pick."""

    index: Annotated[int, Field(description="Candidate index in SAM order.")]
    score: Annotated[float, Field(description="CLIP P(good); 0.0 when pre-filtered.")]
    reason: Annotated[str, Field(description="Prefilter or score tag, e.g. winner / click_miss.")]
    preview_b64: Annotated[str, Field(description="PNG preview with click marker, base64.")]
    clip_crop_b64: Annotated[
        str | None, Field(default=None, description="Gray-composited CLIP crop PNG, or null if not scored.")
    ]
    cutout_b64: Annotated[str, Field(description="BGRA cutout PNG, base64.")]


class DebugAutoMaskPickResponse(BaseModel):
    """All SAM candidates plus CLIP ranking for one click. No session writes."""

    click_xy: Annotated[list[int], Field(description="Natural-image click [x, y].")]
    threshold: Annotated[float, Field(description="Minimum P(good) to become winner.")]
    winner_index: Annotated[int | None, Field(description="Chosen candidate, or null.")]
    candidates: Annotated[list[DebugMaskCandidate], Field(description="Every SAM candidate.")]
    elapsed_ms: Annotated[float, Field(description="Wall time for segment + rank.")]


class DebugSdParams(BaseModel):
    """SD knobs used for one inpaint pass."""

    prompt: Annotated[str, Field(description="Positive prompt.")]
    negative_prompt: Annotated[str, Field(description="Negative prompt.")]
    strength: Annotated[float, Field(description="Denoising strength.")]
    num_inference_steps: Annotated[int, Field(description="Diffusion steps.")]
    guidance_scale: Annotated[float, Field(description="CFG scale.")]
    mask_dilate_pixels: Annotated[int, Field(default=0, description="Verifier mask expansion for next retry.")]
    compose_dilate_pixels: Annotated[int, Field(default=0, description="Verifier compose expansion for next retry.")]


class DebugInpaintAttempt(BaseModel):
    """One verify loop iteration after an SD (or skipped-SD) candidate."""

    attempt_index: Annotated[int, Field(description="0-based verify attempt.")]
    ok: Annotated[bool, Field(description="Whether the verifier passed this candidate.")]
    sd_skipped: Annotated[bool, Field(description="True if this attempt used LaMa only.")]
    scores: Annotated[dict[str, float], Field(description="CLIP label scores when fallback ran.")]
    winner_label: Annotated[str, Field(description="Verifier winner label.")]
    params: Annotated[DebugSdParams, Field(description="Params sent into this SD/verify pass.")]
    param_fixes_json: Annotated[str, Field(description="JSON returned by the verifier.")]
    mask_dilate_pixels: Annotated[int, Field(default=0, description="AI mask dilate for next retry.")]
    compose_dilate_pixels: Annotated[int, Field(default=0, description="AI compose dilate for next retry.")]
    mask_pixel_count: Annotated[int, Field(default=0, description="Inpaint mask pixel count at verify time.")]
    next_params: Annotated[DebugSdParams | None, Field(default=None, description="Parsed retry recipe on fail.")]
    candidate_b64: Annotated[str, Field(description="Full-image candidate PNG, base64.")]
    clip_crop_b64: Annotated[str, Field(description="Padded mask crop sent to verifier, base64.")]


class DebugInpaintVerifyResponse(BaseModel):
    """Hybrid inpaint + CLIP retry trace for one mask. No session writes."""

    click_xy: Annotated[list[int], Field(description="Natural-image click [x, y].")]
    mask_index: Annotated[int, Field(description="Candidate index that was inpainted.")]
    passed: Annotated[bool, Field(description="True if the last attempt passed CLIP.")]
    retries_exhausted: Annotated[bool, Field(description="True if CLIP never passed.")]
    lama_b64: Annotated[str | None, Field(description="LaMa-only PNG, if captured.")]
    final_b64: Annotated[str, Field(description="Final hybrid output after sharpen.")]
    attempts: Annotated[list[DebugInpaintAttempt], Field(description="Verify loop, in order.")]
    elapsed_ms: Annotated[float, Field(description="Wall time for segment + inpaint + verify.")]
